windows/src/engine/keyman32/appint/aiTIP.cpp

```diff
@@ -226,34 +226,52 @@ extern "C" __declspec(dllexport) BOOL WINAPI TIPProcessKey(WPARAM wParam, LPARAM
 	_td->TIPGetContext = ctfunc;
 
   AppContextWithStores *savedContext = NULL;  // I4370   // I4978
-  AppContext *savedContextUsingCore  = NULL; // used for common core
-
+  AppContext *savedContextUsingCore  = NULL;  // used for common core
+  km_kbp_option_item *SavedKBDOptions = NULL; // used for common core
   if (!Updateable) {
     if (isUsingCoreProcessor) {
       savedContextUsingCore = new AppContext();
       _td->app->CopyContext(savedContextUsingCore);
+      SavedKBDOptions = SaveKeyboardOptionsCore(_td->lpActiveKeyboard);
+      // TODO: 5653  ---- Alternate option ---
+      // Instead of allocating memory for each option
+      // just update the changed options. Consider the commented out method UpdateCoreKeyboard
+      // options.
+      /*if (_td->lpActiveKeyboard->lpCoreKeyboardOptions) {
+        UpdateKeyboardOptionsCore(_td->lpActiveKeyboard->lpCoreKeyboardState, _td->lpActiveKeyboard->lpCoreKeyboardOptions);
+      } else {
+        _td->lpActiveKeyboard->lpCoreKeyboardOptions = SaveKeyboardOptionsCore(_td->lpActiveKeyboard);
+      }*/
+      //////// End alternate option
```

Comment:
I am after input as to which is prefered here.  As the comment explains we can store the options away with the thread (cached) and just updated the options that have changed. This will still go through each option on a keypress. Or we can create and destroy memory for the options on each keypress.

Reply:
Either way is probably okay; the cost of freeing and reallocating per keystroke is not a big deal.

Reply:
So I went with the free/allocate each keystroke option.